cpp/src/parquet/arrow/reader.cc

```diff
@@ -967,6 +978,74 @@ Status FileReaderImpl::GetRecordBatchReader(const std::vector<int>& row_groups,
   return Status::OK();
 }
 
+/// Given a file reader and a list of row groups, this is a generator of record
+/// batch vectors (where each vector is the contents of a single row group).
+class RowGroupGenerator {
+ public:
+  using Item = ::arrow::util::optional<::arrow::RecordBatchVector>;
+
+  explicit RowGroupGenerator(FileReaderImpl* self, std::vector<int> row_groups,
+                             std::vector<int> column_indices)
+      : self_(self),
+        index_(0),
+        row_groups_(std::move(row_groups)),
+        column_indices_(std::move(column_indices)) {}
+
+  ::arrow::Future<Item> operator()() {
+    if (index_ >= row_groups_.size()) {
+      return ::arrow::Future<Item>::MakeFinished(::arrow::util::nullopt);
+    }
+    int row_group = row_groups_[index_++];
+    FileReaderImpl* self = self_;
+    std::vector<int> column_indices = column_indices_;
+    ARROW_ASSIGN_OR_RAISE(auto fut,
+                          ::arrow::internal::GetCpuThreadPool()->Submit(
```

Comment:
This will be a problem.  I think this task will block (I'm assuming `ReadRowGroupsImpl` is synchronous and blocking?)  Blocking tasks should not be on the CPU pool.  You could put it on the I/O pool but that isn't necessarily ideal either as it complicates sizing the I/O pool.

Ideally you want something like `reader_->PreBuffer(row_groups, ...).Then(NonBlockingReadRowGroupsImpl)`.

I think you might kind of get away with it because the task they are waiting on is on the I/O thread pool (I assume the prebuffering tasks are on the I/O pool) so you won't have the nested deadlock problem.

However, it will not have ideal performance.  If you are reading a bunch of files you will have some CPU threads tied up waiting that could be doing work.

Reply:
Ah, ok - I was envisioning that the caller would explicitly buffer beforehand (since we had been talking about splitting up ScanTask in that way) but we can have this manage buffering internally as well. (Either way, either the IPC reader or the Parquet reader will need some refactoring to meet Datasets' needs.)

Reply:
Ok, if you plan to do that buffering elsewhere then this is fine.  I guess I misunderstood then.  So the point in breaking things into tasks here is to allow for parallelism?

Reply:
Yes, this is to allow for parallelism when scanning a Parquet file. Though as I think about it, maybe this isn't necessary? We can do the parallelism at the scan task level already.

Reply:
Maybe we can hold this until ARROW-7001 is through and then we can see what exactly we need to be reentrant to get the pipeline we want.